src/main/java/duke/component/DukeDatabase.java

```diff
@@ -0,0 +1,144 @@
+        }
+
+        return this;
+    }
+
+    /**
+     * Returns all the tasks stored in the database.
+     *
+     * @return a list of all the tasks stored in the database.
+     */
+    public TaskList getAllTasks() {
+        TaskList tasksList = new TaskList();
+
+        try {
+            File f = new File(databaseDirectory);
+            Scanner sc = new Scanner(f);
+
+            // Read the database file.
+            while (sc.hasNextLine()) {
+                try {
+                    // Create the corresponding task object from the data in the database
+                    // and add into the taskList.
+                    tasksList.addTask(createTask(sc.nextLine()));
+                } catch (DukeException | NullPointerException | NumberFormatException e) {
+                    System.out.println("Database has corrupted data!");
+                }
+            }
+        } catch (FileNotFoundException e) {
+            System.out.println("Cannot find database file!");
+        }
+
+        return tasksList;
+    }
+
+    /**
+     * Creates the corresponding task object from a line of data in the database.
+     *
+     * @param dataInput a line of data in the database.
+     * @return the corresponding task object.
+     * @throws DukeException If the given data is corrupted.
+     */
+    private Task createTask(String dataInput) throws DukeException {
+        // Get the type of the Task.
+        String[] arr = dataInput.split("\\|");
+        String type = arr[0].trim();
+
+        // Create the corresponding task object.
+        Task task;
+        if ("T".equals(type)) {
+            task = new ToDo(arr[2].trim());
+        } else if ("D".equals(type)) {
+            task = new Deadline(arr[2].trim(), arr[3].trim());
+        } else if ("E".equals(type)) {
+            task = new Event(arr[2].trim(), arr[3].trim());
+        } else {
+            throw new DukeException();
+        }
+
+        // Check and update the status of the task accordingly.
+        int status = Integer.parseInt(arr[1].trim());
+        if (status == 1) {
+            task.markAsDone();
+        } else if (status != 0) {
+            throw new DukeException();
+        }
+
+        return task;
+    }
+
+    /**
+     * Updates the database file (used at the end of the program).
+     *
+     * @param tasks a list of tasks as updated at the point when the program ends.
+     */
+    public void update(TaskList tasks) {
+        // Extract all the data summaries of the tasks and append them into a single string.
+        StringBuilder bldr = new StringBuilder(150);
+        tasks.forEach(t -> {
+            bldr.append(String.format("%s\n", t.getSummaryForDatabase()));
+        });
+
+        // Write all the data summaries into the database file.
+        try {
+            FileWriter fw = new FileWriter(databaseDirectory);
+            fw.write(bldr.toString());
+            fw.close();
```

Comment:
Here it would be advisable to move `fw.close()` into a `finally` block. 

` 
try {
    FileWriter fw = new FileWriter(databaseDirectory);
    fw.write(bldr.toString());
} catch (IOException e) {
    System.out.println("Database error occurs when updating data!");
} finally {
    fw.close();
}
`

The reason for this is, suppose the `try-catch` catches an exception after `fw` has gained file access. When the exception is handled, `fw` still has access to the file in `databaseDirectory`. The next time you try to access this file, you will encounter an IOException again.

The `finally` block always executes regardless of Exceptions caught in the `try` block.